PsfLauncher/ErrorInformation.h

```diff
@@ -0,0 +1,93 @@
+#pragma once
+#include <string>
+#include <windows.h>
+#include <iostream>
+#include "utilities.h"
+
+class ErrorInformation
+{
+public:
+    ErrorInformation()
+    {
+    }
+
+
+    ErrorInformation(std::wstring customMessage, DWORD errorNumber) :
+        ErrorInformation(customMessage, errorNumber, L"")
+    {
+    }
+
+    ErrorInformation(std::wstring customMessage, DWORD errorNumber, std::wstring exeName)
+        : customMessage(customMessage)
+        , errorNumber(errorNumber)
+        , errorMessage(widen(std::system_category().message(errorNumber)))
+        , isThereAnError(true)
+        , exeName(exeName)
+    {
+    }
+
+    ErrorInformation(const ErrorInformation &toCopy)
+    {
+        this->customMessage = toCopy.customMessage;
+        this->errorMessage = toCopy.errorMessage;
+        this->isThereAnError = toCopy.isThereAnError;
+        this->errorNumber = toCopy.errorNumber;
+        this->exeName = toCopy.exeName;
+    }
+
+    ErrorInformation& operator=(const ErrorInformation &toAssign)
+    {
+        if (this == &toAssign)
+        {
+            return *this;
+        }
+
+        this->customMessage.clear();
+        this->customMessage = toAssign.customMessage;
+        this->errorMessage = toAssign.errorMessage;
+        this->isThereAnError = toAssign.isThereAnError;
+        this->errorNumber = toAssign.errorNumber;
+        this->exeName = toAssign.exeName;
+
+        return *this;
+    }
+
+    std::wstring Print()
+    {
+        this->customMessage.clear();
+        this->customMessage.append(L"\r\nMessage: " + this->errorMessage + L"\r\n");
+        this->customMessage.append(L"Error number: " + this->errorNumber);
+        this->customMessage.append(L"\r\n");
+
+        if(!this->exeName.empty())
+        {
+            this->customMessage.append(L"ExeName: " + this->exeName);
+        }
+        this->customMessage.append(L"\r\n");
+
+        return this->customMessage;
+    }
+
+    bool IsThereAnError()
+    {
+        return this->isThereAnError;
+    }
+
+    DWORD GetErrorNumber()
+    {
+        return this->errorNumber;
+    }
+
+    void AddExeName(std::wstring exeNameToAdd)
+    {
+        this->exeName = exeNameToAdd;
+    }
+
+private:
+    std::wstring customMessage;

```

Comment:
>customMessage [](start = 17, length = 13)

still don't understand why this is a member:
values captured by constructors are always overwritten
value returned by Print is never cached

Reply:
You are correct.  customMessage is to hold a custom error message when ErrorInformation was made.  I changed print to build a new std::wstring inside of print.  That way the custom message of the error is preserved.